Fix "ZoneId" bug, add MQTT autoreconnect feature

number42net opened the pull request:
I ran into some issues running both the 2.0.0 and master branches, mainly caused by part of the code expecting the devices dict to contain zoneId and other parts expecting zone_id. Initially I started fixing these by using the dict.get() method, but that was causing the zones to just turn to 0.

This was causing errors like these:

```
2020-10-15 14:13:13 |     | ERROR               | ''zoneId'' on line 725 [Command ZONE_TEMPERATURE, data: '---  I --- 01:180412 --:------ 01:180412 30C9 021 0007C901085402078A030BF804073D05073E0608BB', port: 1]
Traceback (most recent call last):
  File "evogateway.py", line 725, in process_received_message
    COMMANDS[msg.command_code](msg)
  File "evogateway.py", line 875, in zone_temperature
    zone_id = devices[msg.source_id]["zoneId"]
KeyError: 'zoneId'
```

I just noticed that these were also reported in issue #18 

Second minor issue is that the BDR91 is being recognized as a controller instead of a relay, presumably because it's often used as a on/off boiler controller.

number42net committed:
Various bug fixes and adjustments

number42net committed:
Moved mqtt_client.subscribe to mqtt_on_connect() function as recommended in in the Paho MQTT documention. This will take care of automatically resubscribing on reconnect as requested in issue #17

number42net commented:
Also added a fix to automatically resubscribe on reconnect as requested in issue #17

smar000 commented:
Thanks @number42net - much appreciated! I will have a quick look through both the commits later in the day before merging.

smar000 commented:
Just had a quick look but am not sure about the "zoneId" change, as this is the key used in the `devices.json` file, whereas `zone_id` is the python variable used internally within the script. Could you post an extract from your `devices.json` file, to confirm that you are using the same key in your json file.

The use of both `zoneId`/`zone_id` is a bit messy TBH, and sort of evolved from the early days. It is quite possible that there are bugs related to this, but I've been using the script as is for many months and haven't noticed any errors.

If I recall correctly, the `devices` holds the the `devices.json` directly, and so will use "zoneId" as its key (assuming that there has been no change in the `devices.json` file as compared to the sample). Once loaded, all internal collections, such as `zones` should use the `zone_id` for their key. The proper way to get consistency would be to change the key in `devices.json`, but given that it is a breaking change for other users, I've been putting it off. May have to look at it now if it is throwing errors for you all.

smar000 reviewed (changes requested):
The existing `device["zoneId"]` is correct I think, and should remain as is, assuming the `devices.json` file device key format has not been changed. Similarly for `devices[d]["zoneId"]` etc.

Can you pls double check your _devices.json_ file.

"zoneId" is the key used in the _devices.json_ file, whereas `zone_id` is the python variable used internally within the script. 

The use of both `zoneId`/`zone_id` is a bit messy TBH, and sort of evolved from the early days. It is quite possible that there are bugs related to this, but I've been using the script as is for many months and haven't noticed any errors.

If I recall correctly, the `devices` variable holds the the _devices.json_ collection directly, and so will use "zoneId" as its key (assuming that there has been no change in the _devices.json_ file as compared to the sample). Once loaded, all internal collections, such as zones should use the `zone_id` for their key. 

The correct way to get consistency would be to change the key in _devices.json_, but given that it is a breaking change for other users, I've been putting it off. May have to look at it now if it is throwing errors for you all.

Thanks.

smar000 reviewed (commented):
Could you do the mqtt change as a separate PR, so that I can merge it by itself. Many thanks.

smar000 commented on evogateway.py:
```diff
@@ -1210,7 +1210,7 @@ def fault_log(msg):
     elif device_type_id == 0x05:
       device_type =  "DHW"
     elif device_type_id == 0x00:
-      device_type = "CONTROLLER"
```
This is correct as a _controller_ I think. Pls see https://github.com/domoticz/domoticz/blob/851da37f7abb99d0a6060945a48ebd49772f57cc/hardware/EvohomeRadio.cpp#L1759

number42net commented:
Here is a copy of my devices.json:

```{
    "01:180412": {
        "name": "Touch screen",
        "zoneMaster": false,
        "zone_id": 7
    },
    "04:092733": {
        "name": "Bed1",
        "zoneMaster": false,
        "zone_id": 3
    },
    "04:092745": {
        "name": "Office",
        "zoneMaster": false,
        "zone_id": 4
    },
    "04:019193": {
        "name": "Bed2",
        "zoneMaster": false,
        "zone_id": 6
    },
    "04:092747": {
        "name": "Living room",
        "zoneMaster": false,
        "zone_id": 1
    },
    "13:163331": {
        "name": "Underfloor relay",
        "zoneMaster": false,
        "zone_id": 7
    },
    "10:137422":{
        "name": "Boiler",
        "zone_id": -1,
        "zoneMaster": false
    },
    "04:092741": {
        "name": "Bed3",
        "zoneMaster": false,
        "zone_id": 2
    },
    "04:092739": {
        "name": "Wardrobe",
        "zoneMaster": false,
        "zone_id": 5
    },
    "18:056026": {
        "name": "EvoGateway",
        "zoneMaster": true,
        "zone_id": 240
    },
    "04:092763": {
        "name": "Faulty",
        "zoneMaster": false,
        "zone_id": -1
    }
}
```

After going a bit more through the code I realise that probably the problem came from devices.json and devices_new.json using a different tag. My workflow was to let the script discover all my new devices and then just copy and paste the content from devices_new.json into devices.json. And then start making alterations to the names and zone IDs. 

It would be great if you could use the same key in both files like I did in my commit, it doesn't really matter which it is, I can just change my devices.json to match.

As for the controller vs relay, if that's what most other apps expect then I'll just have to live with it as well 🙂

number42net committed:
Changed datetime.now() to datetime.utcnow() where UTC time is implied

The timestamps published to MQTT ends in "Z". As per ISO8601 this indicates a Zero offset from UTC. But datetime.now() returns local time.

number42net committed:
Removed MQTT client.on_disconnect() function

The Paho MQTT client.on_disconnect() function should not be used with client.loop_start() since it prevents automatic reconnects.

number42net commented:
I've added two more commits:

The first changes datetime.now() to datetime.utcnow() when publshing time stampts to MQTT. There is a hardcodes Z at the end of the ISO time string, which implies UTC time. But now() returns local time. I ran into this while writing a client that writes the values from MQTT into InfluxDB.

The second is a follow up on issue #17. While restarting my own MQTT broker I noticed that the issue is not just that the subscriptions are not renewed on reconnect, but that there is not attempt to reconnect at all. I found that this is because there is a on_disconnect handler. After removing this handler the script automatically reconnects as expected.

smar000 commented:
> After going a bit more through the code I realise that probably the problem came from devices.json and devices_new.json using a different tag. My workflow was to let the script discover all my new devices and then just copy and paste the content from devices_new.json into devices.json. And then start making alterations to the names and zone IDs.
> 
> It would be great if you could use the same key in both files like I did in my commit, it doesn't really matter which it is, I can just change my devices.json to match.

Thanks for spotting this. On looking back through the evolution of the code, I can see that the key "zone_id" crept in into the automatically detected new devices file when I did some refactoring some months back (it was originally "zoneId", and consistent with the sample `devices.json` file). The cleanest option is that we keep "zoneId" in the json files and correct the bug introduced in the line writing new devices to file. 

I will push an update tomorrow for this.

smar000 commented:
> I've added two more commits:
> 
> The first changes datetime.now() to datetime.utcnow() when publshing time stampts to MQTT. There is a hardcodes Z at the end of the ISO time string, which implies UTC time. But now() returns local time. I ran into this while writing a client that writes the values from MQTT into InfluxDB.
> 
> The second is a follow up on issue #17. While restarting my own MQTT broker I noticed that the issue is not just that the subscriptions are not renewed on reconnect, but that there is not attempt to reconnect at all. I found that this is because there is a on_disconnect handler. After removing this handler the script automatically reconnects as expected.

Great, thanks for both of these! I will get them merged tomorrow as well after brief testing.

smar000 commented:
The mqtt reconnect and datetime commits have been merged. The "zoneId" matter was addressed separately by correcting the typo in the write line to the `devices_new.json` file.
